### ISSUE TITLE
We shouldn't rely on pstree output to get parent processes

### ISSUE THREAD

**Issue by SeerLite** [Liupold/pidswallow#8]
I was messing around again with `pstree` to understand what `sed` did in the new changes, and this is an example output I got:
```
$ pstree -ATsp "$$" | sed -e 's|(|:|g' -e 's|)||g'
systemd:1---systemd:709---alacritty:2734---zsh:2742-+-pstree:5774
                                                        `-sed:5775
```
I felt like this could break the script, so I started looking around and found [this stackoverflow answer](https://superuser.com/a/1366836/1047393) that recommends not parsing `pstree`. This leads me back to the conclusion that a better solution is to use the `/proc` filesystem once again.
This has the performance problem though, because we have to parse `/proc/<...>/status` for _each and every_ parent until coming across a swallowable.
So right now I'm looking for an alternative to `pstree` that's reliable enough for scripting and know will work 100% of the time.

I know, maybe a scenario like the output I got above will never really happen, but I feel like being nitpicky and really want a solution that always works ;) hehe

**Comment by SeerLite** [comment]
Speaking of... It looks like the proc filesystem is pretty linux-specific.

Now, I'm not completely sure if bspwm is linux-specific or not, but I guess since the script is already pretty POSIX, that `ps` is the only true way to deal with all of this. The commands are
`ps -p <pid> -o comm=` for process name
and
`ps -p <pid> -o ppid=` for parent pid.

Let me know what you think about this!

**Comment by Liupold** [comment]
[pids.with.ps.txt](https://github.com/Liupold/pidswallow/files/4906142/pids.with.ps.txt)
Here is a script with your suggested modification. (around `0.095s`, almost twice as slow). If we left the `pstree` as it is. (The exception you had was a result of parsing the parser). (The `sed` was a pipe and a child to pstree). This will not be there while using this script. (I have tried with pipes it works fine, Eg: `feh`, `mpv`, `zathura`). With only the modification to the `/proc/$cpid/comm -> ps -p $cpid -o comm:1=` it's around `0.050s` (execution). 

* I am going to change `/proc`  to `ps`.
* I am still not sure what to do with pstree. see if you can make any improvement. (1.5x base  is the threshold). [execution time].
* from what i can tell `pstree` is available on `bsd` too.
* yes parsing `pstree` is not ideal. But it is the fastest way (till now).

Any improvement PR will be merged. (feel free to create one, you can always modify your PR and once it's good enough i will merged it.)

**Comment by Liupold** [comment]
I think we can now add `ps` instead of `pstree` as the script now support a `--loop` feature.
(So it doesn't have to execute every line on node_add / node_remove). It will be slower but it will be must less impactful. Give it a try and let me know.

**Comment by SeerLite** [comment]
Hey, nice! The script looks a lot neater! I'll change some stuff I think is important and then maybe look into using `ps`.

> The `sed` was a pipe and a child to pstree

Actually, it was a child of `zsh`. The issue isn't about pipes, it's about subprocesses. For example, if `mpv` creates a subprocess, then the tree will diverge like my example. Still, I'm gonna look into that. Maybe we can keep using `pstree` and just cut out the first line and try to ignore the subprocesses. **Edit:** I actually did this [here](https://github.com/Liupold/pidswallow/pull/9/commits/4a69df285bb3c4bd769f8a85e7ae9a485305521b).

**Comment by SeerLite** [comment]
Closed through 4a69df2
